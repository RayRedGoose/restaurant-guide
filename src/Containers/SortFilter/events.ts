import { FocusEvent } from "react";

export const handleFocus = (e: FocusEvent<HTMLSelectElement>): number =>
  (e.target.size = 6);

export const handleBlur = (e: FocusEvent<HTMLSelectElement>): number =>
  (e.target.size = 1);
