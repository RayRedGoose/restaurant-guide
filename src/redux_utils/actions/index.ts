import {
  IRestaurantObject,
  IAddRestaurantAction
} from 'assets/ts/interfaces';

export const addRestaurants = (restaurants: IRestaurantObject[]): IAddRestaurantAction => ({
  type: 'ADD_RESTAURANTS',
  restaurants: restaurants
});
