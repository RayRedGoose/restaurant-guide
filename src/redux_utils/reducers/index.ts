import { combineReducers } from 'redux';
import restaurants from './restaurants/restaurants';

const rootReducer = combineReducers({
  restaurants
});

export default rootReducer;
