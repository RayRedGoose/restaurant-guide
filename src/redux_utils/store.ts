import { Store, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import rootReducer from "./reducers";
import { IAppStore } from "assets/ts/interfaces";

const store: Store<IAppStore> = createStore(rootReducer, composeWithDevTools());

export default store;
