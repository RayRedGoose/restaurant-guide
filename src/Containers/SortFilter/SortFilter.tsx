import React, { useState, useEffect, ChangeEvent } from "react";
import { useSelector, useDispatch } from "react-redux";
import { IAppStore, sortingTypes } from "assets/ts/interfaces";
import { addSortFilter } from "redux_utils/actions";
import { handleFocus, handleBlur } from "./events";

const SortFilter: React.FC = () => {
  const dispatch = useDispatch();
  const [active, setActive] = useState<sortingTypes>("name");
  const sortFilter = useSelector((store: IAppStore) => store.sortFilter);

  useEffect(() => {
    setActive(sortFilter);
  }, [sortFilter]);

  const toggleFilter = (e: ChangeEvent<HTMLSelectElement>) => {
    e.target.size = 1;
    const selectedValue: string = e.target.value;

    const validatedValue = (): sortingTypes => {
      return selectedValue === "name" ||
        selectedValue === "state" ||
        selectedValue === "name-reverse" ||
        selectedValue === "state-reverse"
        ? selectedValue
        : "name";
    };

    setActive(validatedValue());
    dispatch(addSortFilter(validatedValue()));
  };

  return (
    <section className="filter" id="sorting">
      <select
        onChange={toggleFilter}
        onFocus={handleFocus}
        onBlur={handleBlur}
        value={active}
      >
        <option value="name">name (A-Z)</option>
        <option value="name-reverse">name (Z-A) </option>
        <option value="state">state (A-Z)</option>
        <option value="state-reverse">state (Z-A) </option>
      </select>
    </section>
  );
};

export default SortFilter;
