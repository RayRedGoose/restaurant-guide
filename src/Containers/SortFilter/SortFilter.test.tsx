import React from "react";
import { shallow } from "enzyme";
import SortFilter from "./SortFilter";
import { addSortFilter } from "redux_utils/actions";
import { handleFocus, handleBlur } from "./events";

const mockDispatch = jest.fn();

jest.mock("react-redux", () => ({
  useDispatch: () => mockDispatch,
  useSelector: () => "name",
}));

jest.mock("redux_utils/actions");
jest.mock("./events");

describe("SortFilter", () => {
  let wrapper: any;

  it("should match the snapshot if type is attires", () => {
    wrapper = shallow(<SortFilter />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should call handleFocus on select focus", () => {
    wrapper = shallow(<SortFilter />);
    const defaultEvent = { target: { size: 1 } };

    wrapper.find("select").simulate("focus", defaultEvent);

    expect(handleFocus).toHaveBeenCalled();
  });

  it("should call handleBlur on select unfocus", () => {
    wrapper = shallow(<SortFilter />);
    const defaultEvent = { target: { size: 1 } };

    wrapper.find("select").simulate("blur", defaultEvent);

    expect(handleBlur).toHaveBeenCalled();
  });

  it("should call addSortFilter with value if option is selected", () => {
    const defaultEvent = {
      target: { size: 5, value: "state" },
    };

    wrapper.find("select").simulate("change", defaultEvent);

    expect(addSortFilter).toHaveBeenCalledWith("state");
  });
});
