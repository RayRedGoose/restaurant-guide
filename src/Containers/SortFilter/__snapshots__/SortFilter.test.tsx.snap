// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortFilter should match the snapshot if type is attires 1`] = `
<section
  className="filter"
  id="sorting"
>
  <select
    onBlur={[MockFunction]}
    onChange={[Function]}
    onFocus={[MockFunction]}
    value="name"
  >
    <option
      value="name"
    >
      name (A-Z)
    </option>
    <option
      value="name-reverse"
    >
      name (Z-A) 
    </option>
    <option
      value="state"
    >
      state (A-Z)
    </option>
    <option
      value="state-reverse"
    >
      state (Z-A) 
    </option>
  </select>
</section>
`;
